KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,712 @@
+
+- `registerRetired`: to register retired details. Retired stores the retired address and approvers
+- `removeRetired`: to remove retired details from the array.
+- `registerNewbie`: to register newbie address and its fund distribution.
+- `removeNewbie`: to remove newbie details from the array.
+- `finalizeRegistration`: sets the status to Registered only executed by owner. After this stage, registrations will be restricted.
+- `approve`: to approve a retiredAddress by the admin.
+- `finalizeApproval`: sets the status to Approved. After this stage, approvals will be restricted.
+  </br>
+  **Conditions**
+  - every retiredAddress must be approved
+  - min required admin’s approval is done for retired contract address.
+  - sum of retired’s balance is greater than treasury amount
+    </br>
+- `reset`: resets all storage values to empty objects except rebalanceBlockNumber. It can only be called during Initialization, Registration and Approved status.
+- `finalizeContract`: to record the execution result of the rebalance and finalize the contract. The storage data cannot be modified after this stage.
+
+#### Fallback Function
+
+The smart contract will have a fallback function to revert any payments. Since its a treasury contract it should not accept any payments nor it should allow any withdrawal.
+
+### Core Logic Overview
+
+TBU - describe how to read contract data and execute rebalancing in Klaytn node
+
+## Rationale
+
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+The smart contract is mainly for recording the details, and the Core will execute the fund re-distributions.
+
+### Design decision
+
+#### KLAY transfer is not allowed via smart contracts
+
+As the balance of treasury funds keeps increasing for every block with the block reward its hard to keep track of the balances and rebalance token allocation. So smart contract will only record the rebalanced allocation and the core will execute the allocation reading from the contract.
+
+#### Approval of retiredAddress
+
+To record the addresses in a verifiable manner the addresses are verified in the contract by calling approve method. The retiredAddress can be a Contract address or a Externally Owned Account. If the retired address is a
+
+- EOA : EOA address can be verified when the account holder directly calls the approve function. `msg.sender == retiredAddress`
+- Contract : Contract address can be verified when the admin of the contract calls approve function. The smart contract calls the `getState()` function implemented in the retiredAddress contract to get the admin details. `msg.sender == admin`
```

Comment:
It would be good to define interface with `getState()` to inform the return values of the function.